// react-widgets/lib/index.js
"use strict";

exports.__esModule = true;
exports.setNumberLocalizer = exports.setDateLocalizer = exports.setLocalizers = exports.utils = void 0;

var _configure = _interopRequireDefault(require("./configure"));

var _DropdownList = _interopRequireDefault(require("./DropdownList"));

exports.DropdownList = _DropdownList.default;

var _Combobox = _interopRequireDefault(require("./Combobox"));

exports.Combobox = _Combobox.default;

var _Calendar = _interopRequireDefault(require("./Calendar"));

exports.Calendar = _Calendar.default;

var _DatePicker = _interopRequireDefault(require("./DatePicker"));

exports.DatePicker = _DatePicker.default;

var _TimePicker = _interopRequireDefault(require("./TimePicker"));

exports.TimePicker = _TimePicker.default;

var _DateTimePicker = _interopRequireDefault(require("./DateTimePicker"));

exports.DateTimePicker = _DateTimePicker.default;

var _NumberPicker = _interopRequireDefault(require("./NumberPicker"));

exports.NumberPicker = _NumberPicker.default;

var _Multiselect = _interopRequireDefault(require("./Multiselect"));

exports.Multiselect = _Multiselect.default;

var _SelectList = _interopRequireDefault(require("./SelectList"));

exports.SelectList = _SelectList.default;

var _SlideTransitionGroup = _interopRequireDefault(require("./SlideTransitionGroup"));

var _SlideDownTransition = _interopRequireDefault(require("./SlideDownTransition"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable global-require */
var setLocalizers = _configure.default.setLocalizers,
    setDateLocalizer = _configure.default.setDateLocalizer,
    setNumberLocalizer = _configure.default.setNumberLocalizer;
exports.setNumberLocalizer = setNumberLocalizer;
exports.setDateLocalizer = setDateLocalizer;
exports.setLocalizers = setLocalizers;
var utils = {
  SlideTransitionGroup: _SlideTransitionGroup.default,
  SlideDownTransition: _SlideDownTransition.default
};
exports.utils = utils;
